src/ir/ir_context.h

```diff
@@ -33,6 +33,19 @@ class IRContext {
   IRContext(const IRContext &) = delete;
   IRContext &operator=(const IRContext &) = delete;
 
+  // Create a context with some number of operators and storages pre-seeded.
+  IRContext(std::vector<ir::Operator> operators,
+            std::vector<ir::Storage> storages) {
+    for (auto iter = std::make_move_iterator(operators.begin());
+         iter != std::make_move_iterator(operators.end()); ++iter) {
+      RegisterOperator(std::make_unique<Operator>(*iter));
+    }
```

Comment:
Doesn't the following work here and below?
```suggestion
    for (const auto& op: operators) {
      RegisterOperator(std::move(op));
    }
```

Reply:
It works in the sense that it compiles, but it performs a copy that the current version does not.

While `std::move` will accept a `const` value or a reference, it silently performs a copy in these cases. This is to enable templatized functions to be less annoying if they are instantiated with both `const` and non-`const` types. Because the current code uses a `move` iterator, it can move the values (and maybe copy elide them) into the destination without making a copy.

Reply:
Ah, `const&` was an oversight. Won't removing the `const` ensure it is only moved?
```
for (auto& op: operators) {
      RegisterOperator(std::move(op));
    }
```

Reply:
This is where things get subtle. I think it still does not, but I am not 100% sure. See this: https://abseil.io/tips/77

My count of the above is that, since `op` is a reference pointing into `operators` which is still considered a live name for all of its contents, there are two names at the time when `std::move` is called on `op`. The name `op` is removed, but the name `operators` is still live, so a copy must be performed.

In the case where a move iterator is used, I think it takes care of the two-name situation for you.

But this is getting to the edge of my ability to reason about C++'s strange semantics.

Reply:
another thought, does this work?
```
 for (auto&& op: operators) {
     RegisterOperator(op);
 }
```

Reply:
Your code with move iterators is totally fine. We can leave it as is if none of the options we discussed here work.  I was just trying to see if there is a shorter code to achieve the same thing and I am also learning as I go. I apologize if I come across as being picky. :( 

Based on  https://abseil.io/tips/77, I think the following should work:
```
for (auto& op: operators) {
  RegisterOperator(std::move(op));
}
```
IIUC, std::move creates an rvalue reference (just like a move iterator). Further, RegisterOperator` takes a `unique_ptr`, it can only be moved?

Reply:
That makes sense, I think probing the options is potentially valuable. I'll experiment to see if any of these shorter options work.

Reply:
I tried the above options with `Storage`, which has a `delete`d copy constructor. It turns out that

```C++
for (Storage &storage : storages) {
  RegisterStorage(std::make_unique_ptr(std::move(storage)));
}
```

works as you'd expect! That was a surprise to me, but a welcome one.